File: 0918-maximum-sum-circular-subarray/0918-maximum-sum-circular-subarray.cpp
class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        const int oo = 1e5;
        int minSub = oo;
        int ansMin = oo;
        int maxSub = -oo;
        int ansMax = -oo;
        int s = 0;
        for (int num : nums) {
            minSub = min(num, minSub + num);
            ansMin = min(ansMin, minSub);
            maxSub = max(num, maxSub + num);
            ansMax = max(ansMax, maxSub);
            s += num;
        }
        if (ansMax < 0) {
            return ansMax;
        }
        return max(s - ansMin, ansMax);
    }
};

File: 0918-maximum-sum-circular-subarray/0918-maximum-sum-circular-subarray.py
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        maxSub = -10**5
        maxSum = -10**5
        minSub = 10**5
        minSum = 10**5
        total = 0
        
        for num in nums:
            maxSum = max(num, maxSum + num)
            maxSub = max(maxSub, maxSum)
            minSum = min(num, minSum + num)
            minSub = min(minSub, minSum)
            total += num
            
        return max(maxSub, total - minSub) if maxSub > 0 else maxSub